File: src/main/scala/Chapter3.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.rdd.RDD

object Chapter3 {
  private val spark = SparkSession.builder()
    .appName("Chapter3")
    .master("local[*]")
    .getOrCreate()

  def time_it(largeNumbers: RDD[Int], identifier: String): Unit = {
    val startTime = System.currentTimeMillis

    // First action: Reduce to compute the sum
    val sum = largeNumbers.reduce(_ + _)
    println(s"Sum: $sum")

    // Second action: Count the number of elements
    val count = largeNumbers.count()
    println(s"Count: $count")

    // Calculate the average
    val average = sum / count.toDouble
    println(s"Average: $average")

    val endTime = System.currentTimeMillis
    println(s"[$identifier] Time taken: ${endTime - startTime} ms")
  }

  // Create the base RDD once
  val nums = spark.sparkContext.parallelize(1 to 1_000_000_00)

  // RDDs for with and without caching
  val largeNumbersWithCache = nums.filter(_ > 500000).cache()
  val largeNumbersWithoutCache = nums.filter(_ > 500000)

  // Trigger an action to cache largeNumbersWithCache
  largeNumbersWithCache.count()

  println("========================= With Caching ================================")
  for (_ <- 0 until 5) {
    time_it(largeNumbersWithCache, "With Caching")
    println("Expensive Operation complete\n================================")
  }

  println("====================== Without Caching ================================")
  for (_ <- 0 until 5) {
    time_it(largeNumbersWithoutCache, "Without Caching")
    println("Expensive Operation complete\n================================")
  }

  spark.stop()
}

File: src/main/python/Chapter3.py
from pyspark.sql import SparkSession
import time

# Initialize SparkSession
spark = SparkSession.builder \
    .appName("Chapter3") \
    .master("local[*]") \
    .getOrCreate()

# Function to measure time taken for operations
def time_it(largeNumbers, identifier):
    start_time = time.time()

    # First action: Reduce to compute the sum
    sum_value = largeNumbers.reduce(lambda a, b: a + b)
    print(f"Sum: {sum_value}")

    # Second action: Count the number of elements
    count_value = largeNumbers.count()
    print(f"Count: {count_value}")

    # Calculate the average
    average = sum_value / count_value
    print(f"Average: {average}")

    end_time = time.time()
    print(f"[{identifier}] Time taken: {end_time - start_time} seconds")

# Create the base RDD once
nums = spark.sparkContext.parallelize(range(1, 100000001))

# RDDs for with and without caching
largeNumbersWithCache = nums.filter(lambda x: x > 500000).cache()
largeNumbersWithoutCache = nums.filter(lambda x: x > 500000)

# Trigger an action to cache largeNumbersWithCache
largeNumbersWithCache.count()

print("========================= With Caching ================================")
for _ in range(5):
    time_it(largeNumbersWithCache, "With Caching")
    print("Expensive Operation complete\n================================")

print("====================== Without Caching ================================")
for _ in range(5):
    time_it(largeNumbersWithoutCache, "Without Caching")
    print("Expensive Operation complete\n================================")

# Stop the Spark session
spark.stop()
